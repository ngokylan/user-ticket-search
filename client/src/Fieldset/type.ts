import { ReactNode } from 'react';

export type FieldsetProps = {
  /** Fieldset Enable/Disable status, (optional) */
  isDisabled?: boolean;
  /** Fieldset id attribute, optional id */
  id?: string;
  /** Fieldset class name attribute, (optional) */
  className?: string;
  /** Fieldset content */
  children: ReactNode;
};
