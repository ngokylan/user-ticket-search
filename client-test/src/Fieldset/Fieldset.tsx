import React from 'react';
import { getClass } from '../_lib/helper';
import { FieldsetProps } from './type';
import './Fieldset.scss';

function Fieldset({ id, children, className, isDisabled }: FieldsetProps) {
  return (
    <fieldset
      id={id}
      data-testid={`fieldset-${id || 'default'}`}
      className={getClass('fieldset', className, {
        disabled: isDisabled,
      })}
      disabled={isDisabled}
    >
      {children}
    </fieldset>
  );
}

export default Fieldset;
