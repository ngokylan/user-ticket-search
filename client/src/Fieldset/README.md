Fieldset
```js
import FormItem from '../FormItem';
import Input from '../Input';

<Fieldset>
  <FormItem
    label={'Form Field Label'}
    labelAddon="Option"
    >
    <Input onChange={`onChange`}  autoFocus={true} />
  </FormItem>
</Fieldset>
```

Disabled Fieldset
```js
import FormItem from '../FormItem';
import Input from '../Input';

<Fieldset isDisabled={true}>
  <FormItem
    label={'Form Field Label'}
    labelAddon="Option"
    >
    <Input onChange={`onChange`}  autoFocus={true} />
  </FormItem>
</Fieldset>
```