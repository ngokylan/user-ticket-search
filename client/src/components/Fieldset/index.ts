export { default } from './Fieldset';
export * from './type';
