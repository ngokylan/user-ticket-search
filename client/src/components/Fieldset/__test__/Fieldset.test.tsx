import React from 'react';
import { mount } from 'enzyme';
import Fieldset, { FieldsetProps } from '..';

const defaultProps: FieldsetProps = undefined;
const mountComponent = (customProps: any = {}) => {
  const props = { ...defaultProps, ...customProps };
  const wrapper = mount(<Fieldset {...props} />);
  const component = wrapper.find('fieldset');

  return { props, wrapper, component };
};

describe('Fieldset', () => {
  describe('rendering', () => {
    it('should render component without any crash', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toHaveLength(1);
    });

    it('should be match with snapshot %s', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('props', () => {
    describe('default', () => {
      it('should have expected default props', () => {
        expect(JSON.stringify(defaultProps)).toBe(
          JSON.stringify(Fieldset.defaultProps)
        );
      });

      it.each([
        ['fieldset-default', {}],
        ['fieldset-custom-id', { id: 'custom-id' }],
      ])('should render test ID properly', (expectedTestId, props) => {
        const { component } = mountComponent(props);
        expect(component.prop('data-testid')).toBe(expectedTestId);
      });
    });

    describe('children', () => {
      it('should render children', () => {
        const { wrapper } = mountComponent({ children: 'Hello world' });
        expect(wrapper.text()).toBe('Hello world');
      });
    });

    describe('custom', () => {
      it.each([
        ['id', 'test', { id: 'test' }],
        ['disabled', true, { isDisabled: true }],
        ['disabled', false, { isDisabled: false }],
        ['disabled', undefined, {}],
      ])('should render prop %s as %s', (name, expected, props) => {
        const { component } = mountComponent(props);
        expect(component.prop(name)).toBe(expected);
      });
    });
  });

  describe('interaction', () => {
    it.each([
      ['should', '.fieldset', 1, {}],
      ['should', '.custom-class', 2, { className: 'custom-class' }],
      ['should not', '.fieldset--disabled', 0, {}],
      ['should not', '.fieldset--disabled', 0, { isDisabled: false }],
      ['should', '.fieldset--disabled', 1, { isDisabled: true }],
    ])('%s have %s', (name, element, expected, props) => {
      const { wrapper } = mountComponent(props);
      expect(wrapper.find(element)).toHaveLength(expected);
    });
  });
});
