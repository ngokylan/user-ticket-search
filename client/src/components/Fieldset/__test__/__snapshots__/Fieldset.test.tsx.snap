// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fieldset rendering should be match with snapshot %s 1`] = `
<fieldset
  className="fieldset"
  data-testid="fieldset-default"
/>
`;
